R/tcrDist3Utils.R

```diff
@@ -349,3 +349,295 @@ FormatMetadataForTcrDist3 <- function(metadata,
+#' @param resolution Numeric clustering resolution parameter matching metadata column suffix.
+#' @param annotate_clusters Boolean specifying whether to display clustering information.
+#'
+#' @return Invisibly returns NULL after printing the combined heatmap plot.
+#' @export
+#' @examples
+#' \dontrun{
+#' TCRDistanceHeatmaps(
+#'   seuratObj_TCR = seuratObj,
+#'   assayList = NULL,
+#'   resolution = 0.1,
+#'   annotate_clusters = TRUE
+#' )
+#' }
+TCRDistanceHeatmaps <- function(
+    seuratObj_TCR = NULL,
+    assayList = NULL,
+    resolution = 0.1,
+    annotate_clusters = TRUE
+) {
+  if (is.null(seuratObj_TCR)) {
+    stop("Please provide a Seurat Object with TCR distance assays.")
+  }
+  
+  assays_to_use <- if (is.null(assayList)) {
+    SeuratObject::Assays(seuratObj_TCR)
+  } else {
+    assayList
+  }
+  
+  heatmaps <- list()
+  
+  for (assay in assays_to_use) {
+    distance_matrix <- as.matrix(Seurat::GetAssayData(seuratObj_TCR, assay = assay, layer = "counts"))
+    
+    # Find metadata column from clustering pipeline
+    cluster_column <- paste0("TcrClustR_", assay, "_", resolution)
+    if (!(cluster_column %in% colnames(seuratObj_TCR@meta.data))) {
+      message(paste("Skipping", assay, "- no", cluster_column))
+      next
+    }
+    
+    # Cluster metadata is an array of assay_size x n_assay, so iterate and slice
+    # the correct portion of that array
+    full_cluster_info <- seuratObj_TCR@meta.data[[cluster_column]]
+    n_cells_in_assay <- ncol(distance_matrix)
+    assay_start_index <- 1
+    for (a in SeuratObject::Assays(seuratObj_TCR)) {
+      if (a == assay) break
+      assay_start_index <- assay_start_index + ncol(Seurat::GetAssayData(seuratObj_TCR, assay = a, layer = "counts"))
+    }
+    assay_end_index <- assay_start_index + n_cells_in_assay - 1
+    
+    cluster_info <- full_cluster_info[assay_start_index:assay_end_index]
+    cluster_info <- as.factor(cluster_info)
+    cluster_levels <- levels(cluster_info)
+    cluster_colors <- setNames(
+      if (length(cluster_levels) <= 8) {
+        RColorBrewer::brewer.pal(length(cluster_levels), "Set2")
+      } else if (length(cluster_levels) <= 12) {
+        RColorBrewer::brewer.pal(length(cluster_levels), "Set3")
+      } else {
+        grDevices::colorRampPalette(RColorBrewer::brewer.pal(8, "Set3"))(length(cluster_levels))
+      },
+      cluster_levels
+    )
+    
+    # Get a ComplexHeatmap
+    heatmap_obj <- .TCRDistanceHeatmap(seuratObj_TCR, assay, cluster_info, cluster_colors, annotate_clusters)
+    drawn_heatmap <- draw(heatmap_obj, merge_legend = FALSE, heatmap_legend_side = "right", annotation_legend_side = "right", newpage = FALSE)
+    
+    if (!is.null(drawn_heatmap)) {
+      heatmaps[[assay]] <- drawn_heatmap
+    }
+  }
+  
+  # Composite with patchwork
+  combined_heatmaps <- wrap_plots(lapply(heatmaps, function(hm) grid.grabExpr(draw(hm))), ncol = 1)
```

Comment:
does R throw a fit if you return this? 

I can imagine us wanting to put an alpha heatmap + beta heatmap into a figure panel.

Reply:
Forgot to push this, lol. Returning the patchworks at the end works fine.